### PR TITLE
Add subject property and and value as parameters to jaccard and mat…

### PR DESCRIPTION
**Pull request by Shalsh23**
…rix service

Fixes #485

### CODE REVIEW

**Comment by Shalsh23** [comment]
WIP: tests

**Comment by Shalsh23** [comment]
Test 1:
iris: `["http://purl.obolibrary.org/obo/UBERON_0000151","http://purl.obolibrary.org/obo/UBERON_0000152","http://purl.obolibrary.org/obo/UBERON_0003097"]`
subject_property: `http://purl.obolibrary.org/obo/caro#part_of`
subject_value: `http://purl.obolibrary.org/obo/CARO_0000000`

```
curl -X GET "http://localhost:8082/similarity/jaccard?iris=%5B%22http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0000151%22%2C%22http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0000152%22%2C%22http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0003097%22%5D&subject_property=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2Fcaro%23part_of&subject_value=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FCARO_0000000" -H "accept: application/json"
```
```
{"results":[{"terms":["http://purl.obolibrary.org/obo/UBERON_0000151","http://purl.obolibrary.org/obo/UBERON_0000152"],"score":0.6629213483146067},{"terms":["http://purl.obolibrary.org/obo/UBERON_0000151","http://purl.obolibrary.org/obo/UBERON_0003097"],"score":0.5128205128205128},{"terms":["http://purl.obolibrary.org/obo/UBERON_0000152","http://purl.obolibrary.org/obo/UBERON_0003097"],"score":0.4878048780487805}]}%
```

File: src/main/resources/swaggerDocs/swagger.yaml [hunk 1]
```diff
@@ -879,10 +879,22 @@ paths:
           format: JSON
         - name: path
           in: query
-          description: SPARQL property path composed of full IRIS. This is used to connect the data resource to count (RDF graph world) to the ontology world. E.g. `<http://purl.org/has_state>/<http://purl.org/describes_phenotype>`
+          description: SPARQL property path composed of full IRIS. This is used to connect the data resource to count (RDF graph world) to the ontology world. E.g. `<http://purl.obolibrary.org/obo/RO_0002158>/<http://purl.obolibrary.org/obo/RO_0002487>`
```
Comment (balhoff) [inline src/main/resources/swaggerDocs/swagger.yaml:882]: Why not use a property path that's relevant for our semantic similarity queries? `has_state/describes_phenotype` would be good, with the correct IRIs.

File: src/main/resources/swaggerDocs/swagger.yaml [hunk 2]
```diff
@@ -958,12 +970,24 @@ paths:
...
```
Comment (balhoff) [inline src/main/resources/swaggerDocs/swagger.yaml:973]: Same comment as above.

File: src/main/scala/org/phenoscape/kb/Main.scala [hunk 3]
```diff
@@ -485,12 +501,12 @@ object Main extends HttpApp with App {
                       parameters("terms".as[Seq[IRI]],
                                  "path".as[Path],
                                  "subject_filter_property".as[IRI].?,
-                                 "subject_filter_value".as[IRI].?) { (iris, path, specProp, specVal) =>
-                        validate((specProp.isEmpty && specVal.isEmpty) || (specProp.nonEmpty && specVal.nonEmpty),
+                                 "subject_filter_value".as[IRI].?) { (iris, path, subjProp, subjVal) =>
+                        validate((subjProp.isEmpty && subjVal.isEmpty) || (subjProp.nonEmpty && subjVal.nonEmpty),
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/Main.scala:505]: Can you add this validation to the other two services?

**Review by balhoff — CHANGES_REQUESTED** [review]

File: src/main/scala/org/phenoscape/kb/Main.scala [hunk 4]
```diff
@@ -442,40 +444,66 @@ object Main extends HttpApp with App {
...
-                                   "path".as[Path].?) { (iris, relations, path) =>
+                        formFields(
+                          "iris".as[Seq[IRI]],
+                          "relations".as[Seq[IRI]].?(Seq(rdfsSubClassOf.getIRI, part_of.getIRI)),
+                          "path".as[Path].?,
+                          "subject_property".as[IRI].?,
+                          "subject_value".as[IRI].?
+                        ) { (iris, relations, path, subjProp, subjVal) =>
+                          validate((subjProp.isEmpty && subjVal.isEmpty) || (subjProp.nonEmpty && subjVal.nonEmpty),
+                                   "Subject filter property and value must be provided together if at all")
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/Main.scala:471]: I think this validation needs the same fix.

File: src/main/scala/org/phenoscape/kb/Main.scala [hunk 5]
```diff
@@ -442,40 +444,66 @@ object Main extends HttpApp with App {
...
-                                   "path".as[Path].?) { (terms, relations, path) =>
+                        formFields(
+                          "terms".as[Seq[IRI]],
+                          "relations".as[Seq[IRI]].?(Seq(rdfsSubClassOf.getIRI, part_of.getIRI)),
+                          "path".as[Path].?,
+                          "subject_property".as[IRI].?,
+                          "subject_value".as[IRI].?
+                        ) { (terms, relations, path, subjProp, subjVal) =>
+                          validate((subjProp.isEmpty && subjVal.isEmpty) || (subjProp.nonEmpty && subjVal.nonEmpty),
+                                   "Subject filter property and value must be provided together if at all")
```
Comment (balhoff) [inline src/main/scala/org/phenoscape/kb/Main.scala:504]: This validations needs the fix as well.

**Review by balhoff — CHANGES_REQUESTED** [review]

**Review by balhoff — APPROVED** [review]

**Pull request merged**